### PR TITLE
♻️ Rewrite pages router to enable SSR.

### LINKED ISSUE #2180: Trailing slash woes
I realized that forgetting a trailing slash causes 404, e.g.

Valid: https://amp.dev/documentation/examples/components/amp-accordion/
Invalid: https://amp.dev/documentation/examples/components/amp-accordion

Would be good to make it instead redirect to the correct (canonical) page.

### PR DESCRIPTION
**Pull request by matthiasrohmer**
Fixes #2180 and (build) performance issues related to format-filtering via cheerio.js, foundation for #2295, #1694.

### CODE REVIEW

File: platform/lib/templates/index.js [hunk 1]
```diff
@@ -97,8 +118,14 @@ class Templates {
   async fetchTemplate_(templatePath) {
     const templateUrl = new URL(templatePath, config.hosts.pages.base);
     const fetchResponse = await fetch(templateUrl);
-    return fetchResponse.text();
+
+    if (fetchResponse.status && fetchResponse.status !== 404) {
+      return fetchResponse.text();
+    }
```
Comment (tharders) [inline platform/lib/templates/index.js]: Shouldn't there be a return or an error if the condition does not match

File: platform/lib/templates/index.js [hunk 2]
```diff
@@ -27,11 +27,32 @@ const {pagePath, paths} = require('../utils/project');
...
+ * Builds a Object that adds mandatory variables to the render context
+ * that are needed to successfully SSR the pages
+ * @param  {expressjs.Request} request
+ * @return {Object}
+ */
+function context(request, base) {
+  request = request || {'query': {}};
+  const context = {} || base;
+
+  const ALLOWED_FORMATS = ['websites', 'stories', 'ads', 'email'];
```
Comment (sebastianbenz) [inline platform/lib/templates/index.js]: Templates.js feels like the wrong place for this. Can we move this to the caller?

File: platform/lib/templates/index.js [hunk 3]
```diff
@@ -97,8 +118,14 @@ class Templates {
   async fetchTemplate_(templatePath) {
     const templateUrl = new URL(templatePath, config.hosts.pages.base);
     const fetchResponse = await fetch(templateUrl);
-    return fetchResponse.text();
+
+    if (fetchResponse.status && fetchResponse.status !== 404) {
```
Comment (sebastianbenz) [inline platform/lib/templates/index.js]: use `fetchResponse.ok` instead

File: platform/lib/routers/pages.bak.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
+/**
```
Comment (sebastianbenz) [inline platform/lib/routers/pages.bak.js]: Remove this file as it's stored in the revision history.

File: platform/lib/templates/index.js (hunk 2 above)
Comment (matthiasrohmer) [reply platform/lib/templates/index.js]: My reasoning was that potentially any template that is loaded here could rely on having `format` defined. If we move it to the caller we'd need to define it in both: 

https://github.com/ampproject/docs/blob/7c944d4f14e44d8e74282e117a086ed96931edea/examples/source/1.components/amp-live-list/api.js#L22

and 

https://github.com/ampproject/docs/blob/7c944d4f14e44d8e74282e117a086ed96931edea/platform/lib/routers/pages.js#L162

File: platform/lib/templates/index.js (hunk 3 above)
Comment (matthiasrohmer) [reply platform/lib/templates/index.js]: `Response.ok` would be `false` if Grow returned a error debug page which would have a 500 status but you'd like to return to the user. That's why I went for the more verbose status checking.

Comment (sebastianbenz) [reply platform/lib/templates/index.js]: add a comment in that case

**Comment by matthiasrohmer** [comment]
@sebastianbenz, accomodated your requested changes, only waiting for feedback on one. Additionally  I made some small last changes: when ensuring the URL scheme the redirects are now happening by 301s instead of 302s, all tests pass now and the canonical URL is correct.

File: pages/content/amp-dev/documentation/guides-and-tutorials/index.html [hunk 5]
```diff
@@ -1,26 +1,41 @@
...
 $titles:
   navigation: Guides & Tutorials
 $view: /views/overview/guides.j2
-$path: /documentation/guides-and-tutorials/index.html
-$localization:
-  path: /{locale}/documentation/guides-and-tutorials/index.html
 $order: 1
 $hidden: true
-description: 'All resources you need to build stunning websites with AMP: Guides, tutorials, reference documentation and examples.'
+description: 'All resources you need to build stunning [[ format ]] with AMP: Guides, tutorials, reference documentation and examples.'
```
Comment (sebastianbenz) [inline pages/content/amp-dev/documentation/guides-and-tutorials/index.html]: This won't work for email: `build stunning email` with AMP

File: pages/content/amp-dev/documentation/guides-and-tutorials/index.html [hunk 6]
```diff
@@ -1,26 +1,41 @@
...
 flyout:
   description: Get started with AMP
   icon: guides-documents
 ---
 <section class="ap--section">
   <h1>{{ doc.title }}</h1>
-  <p>Here you can find everything you need to build AMP experiences. The sidebar on the left shows you all resources for building <strong>AMP websites.</strong></p>
+  <p>Here you can find everything you need to build AMP experiences. The sidebar on the left shows you all resources for building <strong>AMP [[ format ]].</strong></p>
```
Comment (sebastianbenz) [inline pages/content/amp-dev/documentation/guides-and-tutorials/index.html]: same as above

**Comment by matthiasrohmer** [comment]
Thanks for the review, @sebastianbenz! Especially a good catch with the email plural thing.

**Comment by matthiasrohmer** [comment]
Something I thought about: should we use LRU here as well over a Map, as the later one would need to hold 2GB in the worst case.

https://github.com/ampproject/docs/blob/dd2b357d8aaee6ed25413953783b5a7498a195da/platform/lib/templates/index.js#L78

**Comment by sebastianbenz** [comment]
@matthiasrohmer good catch, we definitely should use an LRU cache here (not sure what a good cache size would be though). It's ironic that we now run into these kinds of problems we wouldn't have with a purely server-side-rendered or static site...

**Comment by matthiasrohmer** [comment]
Ran a test build to check: one locale with 862 pages weighs 82MB in total, making 984MB in total for all locales. Our VMs seem to be n1-standard-2 which have a total of 7,5GB memory though Node.js won't get it all and might be [limited at maybe 1,4GB]. 

So I assume fitting three full locales (246MB) into the LRU will not cause any problem. That would mean capping it at 2586 entries. WDYT?

**Comment by matthiasrohmer** [comment]
@sebastianbenz, please take another look.

**Comment by matthiasrohmer** [comment]
Added another context variable `forceFiltered` that is true if no valid format is given as GET parameter. That's needed for the tools page for example to be able to show all tools if no explicit format is given.

File: pages/content/amp-dev/documentation/guides-and-tutorials/index.html [hunk 7]
```diff
@@ -1,26 +1,41 @@
...
 $titles:
   navigation: Guides & Tutorials
 $view: /views/overview/guides.j2
-$path: /documentation/guides-and-tutorials/index.html
-$localization:
-  path: /{locale}/documentation/guides-and-tutorials/index.html
 $order: 1
 $hidden: true
-description: 'All resources you need to build stunning websites with AMP: Guides, tutorials, reference documentation and examples.'
+description: "All resources you need to build stunning [[ format if format !== 'email' else 'emails' ]] with AMP: Guides, tutorials, reference documentation and examples."
```
Comment (sebastianbenz) [inline pages/content/amp-dev/documentation/guides-and-tutorials/index.html]: Maybe use a helper `pluralize` instead.

Comment (sebastianbenz) [reply pages/content/amp-dev/documentation/guides-and-tutorials/index.html]: s/build/create

File: platform/lib/routers/pages.js [hunk 8]
```diff
@@ -17,173 +17,169 @@
...
-  });
+  // As the request path is not the actual path to the template it is somehow
+  // guessed by testing all of AVAILABLE_STUBS ...
+  for (const stub of AVAILABLE_STUBS) {
+    // Othwerwise try the first stub or the already resolved path if there is one
+    const searchPath = resolvedPath || `${templatePath}${stub}`;
+    try {
+      template = await Templates.get(searchPath);
+    } catch (e) {
+      continue;
```
Comment (sebastianbenz) [inline platform/lib/routers/pages.js]: add a comment why this is expected.

File: platform/lib/routers/pages.js [hunk 9]
```diff
@@ -17,173 +17,169 @@
...
 
-  // Also create a logger during development since you want to know
-  // what's going on
-  const log = new Signale({
-    'interactive': false,
-    'scope': 'Grow (Proxy)',
-  });
+  // As the request path is not the actual path to the template it is somehow
+  // guessed by testing all of AVAILABLE_STUBS ...
+  for (const stub of AVAILABLE_STUBS) {
```
Comment (sebastianbenz) [inline platform/lib/routers/pages.js]: Moving this part into a method makes it easy to implement this without `continue` and `break` making it a lot easier to follow the control flow.

**Pull request merged**